xarray/core/indexing.py

```diff
@@ -18,7 +31,148 @@
     is_duck_dask_array,
     sparse_array_type,
 )
-from .utils import maybe_cast_to_coords_dtype
+from .utils import either_dict_or_kwargs
+
+if TYPE_CHECKING:
+    from .dataarray import DataArray
+    from .dataset import Dataset
+    from .indexes import Index, IndexVars
+
+
+@dataclass
+class QueryResult:
+    """Index query results.
+
+    Attributes
+    ----------
+    dim_indexers: dict
+        A dictionary where keys are array dimensions and values are
+        location-based indexers.
+    indexes: dict, optional
+        New indexes to replace in the resulting DataArray or Dataset.
+    index_vars : dict, optional
+        New indexed variables to replace in the resulting DataArray or Dataset.
+    drop_coords : list, optional
+        Coordinate(s) to drop in the resulting DataArray or Dataset.
+    rename_dims : dict, optional
+        A dictionnary in the form ``{old_dim: new_dim}`` for dimension(s) to
+        rename in the resulting DataArray or Dataset.
+
+    """
+
+    dim_indexers: Dict[Any, Any]
+    indexes: Dict[Hashable, "Index"] = field(default_factory=dict)
+    index_vars: "IndexVars" = field(default_factory=dict)
+    drop_coords: List[Hashable] = field(default_factory=list)
+    rename_dims: Dict[Any, Hashable] = field(default_factory=dict)
+
+
+def merge_query_results(results: List[QueryResult]) -> QueryResult:
+    all_dims_count = Counter([dim for res in results for dim in res.dim_indexers])
+    duplicate_dims = {k: v for k, v in all_dims_count.items() if v > 1}
+
+    if duplicate_dims:
+        fmt_dims = [
+            f"{dim!r}: {count} indexes involved"
+            for dim, count in duplicate_dims.items()
+        ]
+        raise ValueError(
+            "Xarray does not support label-based selection with more than one index"
+            "over the following dimension(s):\n"
+            + "\n".join(fmt_dims)
+            + "Suggestion: use a multi-index for each of those dimension(s)."
+        )
+
+    dim_indexers = {}
+    indexes = {}
+    index_vars = {}
+    drop_coords = []
+    rename_dims = {}
+
+    for res in results:
+        dim_indexers.update(res.dim_indexers)
+        indexes.update(res.indexes)
+        index_vars.update(res.index_vars)
+        drop_coords += res.drop_coords
+        rename_dims.update(res.rename_dims)
+
+    return QueryResult(dim_indexers, indexes, index_vars, drop_coords, rename_dims)
+
+
+def group_indexers_by_index(
+    obj: Union["DataArray", "Dataset"],
```

Comment:
```suggestion
    obj: T_Xarray,
```
You can import that from https://github.com/pydata/xarray/blob/7bfee3eaa8fd731494cf6b406d6abb4bec061001/xarray/core/types.py#L24